### PR TITLE
[AIR] Update Torch benchmarks with documentation

### PR DESCRIPTION
**Pull request by jiaodong**
## Why are these changes needed?

* Used ImageFolderDatasource to simplify image reading 
* Changed pytorch default epoch for test latency with configurable workers
* Added benchmark metrics to benchmark.rst

## Issues found

There're a few more perf issues found during batch prediction / training benchmarks.

1) Batch prediction seems to have very significant overhead compare to training, which supposedly did forward + backward passes compare to prediction that's forward only. **I suspect the in-line conversions in TorchPredictor had something to do with it.**
2) Trainer didn't seem to scale well in a cluster of GPUs. In my setup I had 16 workers, each taking one GPU; However the observed performance is very slow with GPU utils at 0% for 30+ mins .. suggesting we're probably taking heavy perf hit handing data before training even starts. But the same scale of data (100GB) with same cluster setup (8 x g3.8xlarge) worked well for GPU batch predictions.
    30mins+ in the spilling is pretty bad
```
======== Object references status: 2022-07-16 15:05:51.607174 ========
--- Aggregate object store stats across all nodes ---
Plasma memory usage 97950 MiB, 2950 objects, 16.65% full, 16.51% needed
Spilled 873271 MiB, 1519 objects, avg write throughput 2554 MiB/s
Restored 90147 MiB, 53 objects, avg read throughput 7210 MiB/s
Objects consumed by Ray tasks: 876708 MiB.
```

Screen Shot 2022-07-16 at 3 26 21 PM

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]

File: doc/source/ray-air/benchmarks.rst [hunk 1]
```diff
@@ -131,3 +131,82 @@ XGBoost parameters were kept as defaults for xgboost==1.6.1 this task.
 .. _`XGBoost Training Script`: https://github.com/ray-project/ray/blob/a241e6a0f5a630d6ed5b84cce30c51963834d15b/release/air_tests/air_benchmarks/workloads/xgboost_benchmark.py#L40-L58
 .. _`XGBoost Prediction Script`: https://github.com/ray-project/ray/blob/a241e6a0f5a630d6ed5b84cce30c51963834d15b/release/air_tests/air_benchmarks/workloads/xgboost_benchmark.py#L63-L71
 .. _`XGBoost Cluster Configuration`: https://github.com/ray-project/ray/blob/a241e6a0f5a630d6ed5b84cce30c51963834d15b/release/air_tests/air_benchmarks/xgboost_compute_tpl.yaml#L6-L24
+.. _`GPU image batch prediction script`: https://github.com/ray-project/ray/blob/cec82a1ced631525a4d115e4dc0c283fa4275a7f/release/air_tests/air_benchmarks/workloads/gpu_batch_prediction.py#L18-L49
+.. _`GPU image training script`: https://github.com/ray-project/ray/blob/cec82a1ced631525a4d115e4dc0c283fa4275a7f/release/air_tests/air_benchmarks/workloads/pytorch_training_e2e.py#L95-L106
+
```
Comment (richardliaw) [inline doc/source/ray-air/benchmarks.rst:136]: can you move all below sections above the hyperlinks?

File: doc/source/ray-air/benchmarks.rst [hunk 2]
```diff
@@ -131,3 +131,82 @@ XGBoost parameters were kept as defaults for xgboost==1.6.1 this task.
...
+GPU image training
+------------------------
+
+This task uses the TorchTrainer module to train different amounts of data
+using an Pytorch ResNet model.
+
+We test out the performance across different cluster sizes and data sizes.
+
+- `GPU image batch prediction script`_
```
Comment (richardliaw) [inline doc/source/ray-air/benchmarks.rst:184]: should link to `training` script right?

File: doc/source/ray-air/benchmarks.rst [hunk 3]
```diff
@@ -131,3 +131,82 @@ XGBoost parameters were kept as defaults for xgboost==1.6.1 this task.
...
+      - 40.7 images/sec
+      - `python pytorch_training_e2e.py --data-size-gb=1`
+    * - 1 g3.8xlarge node
+      - 1 worker
+      - 2 epochs
+      - 20 GB
+      - 32460 images
+      - 1388.33 s
+      - 46.76 images/sec
+      - `python pytorch_training_e2e.py --data-size-gb=20`
```
Comment (richardliaw) [inline doc/source/ray-air/benchmarks.rst:212]: hmm, we didn't go to multiple workers?

Comment (jiaodong) [reply doc/source/ray-air/benchmarks.rst:212]: i tried like ~5 times across a week but never had a single run that can complete without spilling huge amount data yet ... in fact I had one running on 8 hosts this morning with 16 GPUs and it's still running =.=

File: doc/source/ray-air/benchmarks.rst (hunk 2 above)
Comment (jiaodong) [reply doc/source/ray-air/benchmarks.rst:184]: good catch !

File: doc/source/ray-air/benchmarks.rst (hunk 3 above)
Comment (jiaodong) [reply doc/source/ray-air/benchmarks.rst:212]: I can update this part once it's done, but if we don't want to look too terrible for training at scale we probably need to try this with larger memory hosts ..

Comment (richardliaw) [reply doc/source/ray-air/benchmarks.rst:212]: what about 1 host 4 workers?

Comment (jiaodong) [reply doc/source/ray-air/benchmarks.rst:203]: i think that could work with 20GB data. 95% sure disk spilling will ruin everything on 100GB.

File: doc/source/ray-air/benchmarks.rst [hunk 4]
```diff
@@ -126,8 +126,87 @@ XGBoost parameters were kept as defaults for xgboost==1.6.1 this task.
...
+This task uses the BatchPredictor module to process different amounts of data
+using a Pytorch pre-trained ResNet model.
+
+We test out the performance across different cluster sizes and data sizes.
+
+- `GPU image batch prediction script`_
+
+.. list-table::
+
+    * - **Cluster Setup**
```
Comment (ericl) [inline doc/source/ray-air/benchmarks.rst:141]: Let's condense the table into 4 rows (as in this PR: https://github.com/ray-project/ray/pull/26633)
- Cluster Setup: 1 g3.8xlarge node
- Data Size: 1 GB (1623 images)
- Performance: 72.59 s (22.3 images/s)
- Command: ...

**Comment by ericl** [comment]
> Batch prediction seems to have very significant overhead compare to training, which supposedly did forward + backward passes compare to prediction that's forward only. I suspect the in-line conversions in TorchPredictor had something to do with it.

Yep, we should update after https://github.com/ray-project/ray/pull/26616 is merged, which addresses this issue (both in bulk and pipelined modes).

> 30mins+ in the spilling is pretty bad

Will be addressed with pipelining in the future (should be usable after my current set of PRs is merged).

File: doc/source/ray-air/benchmarks.rst (hunk 3 above)
Comment (jiaodong) [reply doc/source/ray-air/benchmarks.rst:203]: let me try a 4 - g3.16xlarge setup.

Comment (richardliaw) [reply doc/source/ray-air/benchmarks.rst:203]: yeah, 1 g3.16xlarge with 4 workers.

**Comment by jiaodong** [comment]
> > Batch prediction seems to have very significant overhead compare to training, which supposedly did forward + backward passes compare to prediction that's forward only. I suspect the in-line conversions in TorchPredictor had something to do with it.
> 
> Yep, we should update after #26616 is merged, which addresses this issue (both in bulk and pipelined modes).
> 
> > 30mins+ in the spilling is pretty bad
> 
> Will be addressed with pipelining in the future (should be usable after my current set of PRs is merged).

Great ! Looking forward to review them once they're ready. I didn't quite understand the huge discrepancy of data spilling behavior between `TorchPredictor` and `TorchTrainer` however, in both benchmarks I'm using exactly the same ray cluster setup (8 x g3.8xlarge), exactly the same processor code (except training appends a label as list of int), but the observed perf difference is more than 3x.

Unless the appended label column added significant memory overhead .. did we have different data processing / handling logic between trainer and predictor ?

File: doc/source/ray-air/benchmarks.rst (hunk 3 above)
Comment (ericl) [reply doc/source/ray-air/benchmarks.rst:203]: cc @matthewdeng

File: doc/source/ray-air/benchmarks.rst (hunk 4 above)
Comment (richardliaw) [reply doc/source/ray-air/benchmarks.rst:141]: fixed!
